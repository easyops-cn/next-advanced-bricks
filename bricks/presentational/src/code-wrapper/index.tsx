import React, {
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { createDecorators } from "@next-core/element";
import { ReactNextElement, wrapBrick } from "@next-core/react-element";
import { unwrapProvider } from "@next-core/utils/general";
import type { Button, ButtonProps } from "@next-bricks/basic/button";
import type { copyToClipboard as _copyToClipboard } from "@next-bricks/basic/data-providers/copy-to-clipboard";
import classNames from "classnames";
import "@next-core/theme";
import styleText from "./styles.shadow.css";

const COPY_BUTTON_ICON = {
  lib: "fa",
  prefix: "far",
  icon: "copy",
} as const;

const COPY_BUTTON_ICON_SUCCESS = {
  lib: "fa",
  prefix: "fas",
  icon: "check",
} as const;

const COPY_BUTTON_ICON_FAILED = {
  lib: "fa",
  prefix: "fas",
  icon: "triangle-exclamation",
} as const;

export const WrappedButton = wrapBrick<Button, ButtonProps>("eo-button");

export const copyToClipboard = unwrapProvider<typeof _copyToClipboard>(
  "basic.copy-to-clipboard"
);

const { defineElement, property } = createDecorators();

export interface CodeWrapperProps {
  preProps?: JSX.IntrinsicElements["pre"];
  showCopyButton?: boolean;
  themeVariant?: "default" | "elevo";
}

/**
 * 构件 `presentational.code-wrapper`
 *
 * @internal
 *
 * @part pre - 包裹代码内容的 `<pre>` 元素
 * @part copy - 复制按钮
 */
export
@defineElement("presentational.code-wrapper", {
  styleTexts: [styleText],
})
class CodeWrapper extends ReactNextElement implements CodeWrapperProps {
  @property({ attribute: false })
  accessor preProps:
    | React.PropsWithChildren<JSX.IntrinsicElements["pre"]>
    | undefined;

  /**
   * 是否展示复制按钮
   * @default true
   */
  @property({ type: Boolean })
  accessor showCopyButton: boolean | undefined;

  /**
   * 主题变体
   */
  @property()
  accessor themeVariant: "default" | "elevo" | undefined;

  render() {
    return (
      <CodeWrapperComponent
        preProps={this.preProps}
        showCopyButton={this.showCopyButton}
        themeVariant={this.themeVariant}
      />
    );
  }
}

function CodeWrapperComponent({
  preProps,
  showCopyButton,
  themeVariant,
}: CodeWrapperProps) {
  const ref = useRef<HTMLPreElement>(null);
  const [state, setState] = useState<"idle" | "success" | "failed">("idle");
  const timerRef = useRef<ReturnType<typeof setTimeout> | undefined>();

  const handleCopy = useCallback(async () => {
    clearTimeout(timerRef.current);
    setState("idle");
    const content = ref.current?.querySelector("code")?.textContent;
    if (content) {
      try {
        await copyToClipboard(content);
        setState("success");
      } catch (error) {
        // eslint-disable-next-line no-console
        console.error("Failed to copy code block content to clipboard.", error);
        setState("failed");
      }

      timerRef.current = setTimeout(() => {
        setState("idle");
      }, 2000);
    }
  }, []);

  useEffect(() => {
    return () => {
      clearTimeout(timerRef.current);
    };
  }, []);

  const { children, ...restProps } = preProps ?? {};

  const prefParts = useMemo(() => {
    const parts = ["pre"];
    if (restProps.className?.includes("mermaid")) {
      parts.push("mermaid");
    }
    return parts.join(" ");
  }, [restProps.className]);

  return (
    <pre {...restProps} part={prefParts} ref={ref}>
      {children}
      {showCopyButton !== false && (
        <WrappedButton
          part="copy"
          themeVariant={themeVariant}
          className={classNames("copy", {
            success: state === "success",
            failed: state === "failed",
          })}
          onClick={handleCopy}
          icon={
            state === "success"
              ? COPY_BUTTON_ICON_SUCCESS
              : state === "failed"
                ? COPY_BUTTON_ICON_FAILED
                : COPY_BUTTON_ICON
          }
        />
      )}
    </pre>
  );
}
