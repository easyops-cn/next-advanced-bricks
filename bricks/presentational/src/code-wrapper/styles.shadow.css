:host {
  display: block;
  position: relative;
  margin: 0.5em 0;
}

:host([hidden]) {
  display: none;
}

* {
  box-sizing: border-box;
}

pre {
  font-family:
    "Fira Code", "Fira Mono", Menlo, Consolas, "DejaVu Sans Mono", monospace;
  white-space: pre;
  word-spacing: normal;
  word-break: normal;
  line-height: 1.5;
  tab-size: 4;
  hyphens: none;
  font-size: 13px;
  padding: 1em;
  margin: 0;
  overflow-x: auto;
  border-radius: 9px;
}

pre:not(.mermaid) {
  border: 1px solid var(--theme-gray-border-color);
}

pre.mermaid {
  background: var(--palette-gray-3);
}

.copy {
  position: absolute;
  top: 8px;
  right: 8px;
}

:host(:not(:hover)) .copy {
  display: none;
}

.success::part(button) {
  color: var(--color-success);
}

.failed::part(button) {
  color: var(--color-error);
}

.mermaid {
  display: flex;
  justify-content: center;
}
