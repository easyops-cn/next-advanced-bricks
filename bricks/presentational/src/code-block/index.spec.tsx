import { describe, test, expect, jest } from "@jest/globals";
import { act } from "react-dom/test-utils";
import { waitFor } from "@testing-library/react";
import "./";
import "../code-wrapper";
import type { CodeBlock } from "./index.js";

jest.mock("@next-core/theme", () => ({}));

describe("presentational.code-block", () => {
  test("basic usage", async () => {
    const element = document.createElement(
      "presentational.code-block"
    ) as CodeBlock;
    element.language = "js";
    element.source = "console.log('Hello');";

    act(() => {
      document.body.appendChild(element);
    });
    expect(element.shadowRoot?.childNodes).toMatchInlineSnapshot(`
      NodeList [
        <style>
          styles.shadow.css
        </style>,
        <eo-icon
          class="loading"
          icon="loader-circle"
          lib="lucide"
          spinning=""
        />,
      ]
    `);

    await waitFor(
      () => {
        expect(
          element.shadowRoot?.querySelector("presentational\\.code-wrapper")
        ).toBeTruthy();
      },
      {
        timeout: 3000,
      }
    );

    expect(
      element.shadowRoot?.querySelector("presentational\\.code-wrapper")
        ?.shadowRoot.childNodes
    ).toMatchSnapshot();

    act(() => {
      document.body.removeChild(element);
    });
  });

  test("mermaid in markdown", async () => {
    const element = document.createElement(
      "presentational.code-block"
    ) as CodeBlock;
    element.language = "md";
    element.source =
      "```mermaid\nsequenceDiagram\n    Alice->>John: Hello John, how are you?\n    John-->>Alice: Great!\n```";
    element.theme = "dark-plus";
    element.themeVariant = "elevo";
    element.showCopyButton = false;

    expect(element.shadowRoot).toBeFalsy();

    act(() => {
      document.body.appendChild(element);
    });

    await waitFor(
      () => {
        expect(
          element.shadowRoot?.querySelector("presentational\\.code-wrapper")
        ).toBeTruthy();
      },
      {
        timeout: 3000,
      }
    );

    expect(
      element.shadowRoot?.querySelector("presentational\\.code-wrapper")
        ?.shadowRoot.childNodes
    ).toMatchSnapshot();

    act(() => {
      document.body.removeChild(element);
    });
  });
});
