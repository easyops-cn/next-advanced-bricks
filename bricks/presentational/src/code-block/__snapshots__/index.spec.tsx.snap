// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`presentational.code-block basic usage 2`] = `
NodeList [
  <style>
    styles.shadow.css
  </style>,
  <pre
    class="shiki light-plus"
    part="pre"
    style="background-color: rgb(255, 255, 255); color: rgb(0, 0, 0);"
    tabindex="0"
  >
    <code>
      <span
        class="line"
      >
        <span
          style="color: rgb(0, 16, 128);"
        >
          console
        </span>
        <span
          style="color: rgb(0, 0, 0);"
        >
          .
        </span>
        <span
          style="color: rgb(121, 94, 38);"
        >
          log
        </span>
        <span
          style="color: rgb(0, 0, 0);"
        >
          (
        </span>
        <span
          style="color: rgb(163, 21, 21);"
        >
          'Hello'
        </span>
        <span
          style="color: rgb(0, 0, 0);"
        >
          );
        </span>
      </span>
    </code>
    <eo-button
      class="copy"
      icon="[object Object]"
      part="copy"
    />
  </pre>,
]
`;

exports[`presentational.code-block mermaid in markdown 1`] = `
NodeList [
  <style>
    styles.shadow.css
  </style>,
  <pre
    class="shiki dark-plus"
    part="pre"
    style="background-color: rgb(30, 30, 30); color: rgb(212, 212, 212);"
    tabindex="0"
  >
    <code>
      <span
        class="line"
      >
        <span
          style="color: rgb(212, 212, 212);"
        >
          \`\`\`mermaid
        </span>
      </span>
      

      <span
        class="line"
      >
        <span
          style="color: rgb(197, 134, 192);"
        >
          sequenceDiagram
        </span>
      </span>
      

      <span
        class="line"
      >
        <span
          style="color: rgb(156, 220, 254);"
        >
              Alice
        </span>
        <span
          style="color: rgb(197, 134, 192);"
        >
          -&gt;&gt;
        </span>
        <span
          style="color: rgb(156, 220, 254);"
        >
          John
        </span>
        <span
          style="color: rgb(197, 134, 192);"
        >
          :
        </span>
        <span
          style="color: rgb(206, 145, 120);"
        >
           Hello John, how are you?
        </span>
      </span>
      

      <span
        class="line"
      >
        <span
          style="color: rgb(156, 220, 254);"
        >
              John
        </span>
        <span
          style="color: rgb(197, 134, 192);"
        >
          --&gt;&gt;
        </span>
        <span
          style="color: rgb(156, 220, 254);"
        >
          Alice
        </span>
        <span
          style="color: rgb(197, 134, 192);"
        >
          :
        </span>
        <span
          style="color: rgb(206, 145, 120);"
        >
           Great!
        </span>
      </span>
      

      <span
        class="line"
      >
        <span
          style="color: rgb(212, 212, 212);"
        >
          \`\`\`
        </span>
      </span>
    </code>
  </pre>,
]
`;
